frontend/src/pages/new_question/utils/queryBuilderUtils.js

```diff
@@ -0,0 +1,199 @@
+ * @param {object} q_graph valid query graph with nodes and edges
+ * @returns {string} the id of the root node in the query graph
+ */
+function findRootNode(q_graph) {
+  // create nodes object with pinned boolean property
+  const nodes = Object.entries(q_graph.nodes).map(([key, node]) => (
+    {
+      key,
+      pinned: node.id && Array.isArray(node.id) && node.id.length > 0,
+    }
+  ));
+  // create node map of total edge connections
+  const edgeNums = getNumAttachedEdges(q_graph);
+  // split nodes into pinned and unpinned arrays
+  const unpinnedNodes = nodes.filter((node) => !node.pinned && node.key in edgeNums);
+  const pinnedNodes = nodes.filter((node) => node.pinned && node.key in edgeNums);
+  // sort nodes by edge connections, then return the first one in the list
+  // is also first one inserted
+  let root = null;
+  if (unpinnedNodes.length) {
+    unpinnedNodes.sort((a, b) => edgeNums[b.key] - edgeNums[a.key]);
+    root = unpinnedNodes[0].key;
+  } else if (pinnedNodes.length) {
+    pinnedNodes.sort((a, b) => edgeNums[b.key] - edgeNums[a.key]);
+    root = pinnedNodes[0].key;
+  }
+  return root;
+}
+
+/**
+ * Remove all detached sections of graph from root node
+ * @param {object} q_graph deep copy of the query_graph, modifies and returns
+ * @param {string} rootNode root node of graph
+ */
+function trimDetached(q_graph, rootNode) {
+  // all edges start out as disconnected and we'll remove them when we find a connection to root
+  const disconnectedEdges = new Set(Object.keys(q_graph.edges));
+  // we add to connected nodes, starting default with the root node
+  const connectedNodes = new Set([rootNode]);
+  let foundConnections = true;
+  // traverse graph and delete all edges not connected
+  while (foundConnections) {
+    let foundConnection = false;
+    // loop over all connected nodes to find edges that are connected to them
+    [...connectedNodes].forEach((nodeId) => {
+      [...disconnectedEdges].forEach((e) => {
+        const edge = q_graph.edges[e];
+        // if edge is connected to node
+        if (edge.subject === nodeId || edge.object === nodeId) {
+          connectedNodes.add(edge.subject);
+          connectedNodes.add(edge.object);
+          disconnectedEdges.delete(e);
+          foundConnection = true;
+        }
+      });
+    });
+    // no more connected edges, break out
+    if (!foundConnection) {
+      foundConnections = false;
+    }
+  }
+  q_graph.edges = _.omitBy(q_graph.edges, (e, id) => disconnectedEdges.has(id));
+  // delete all floating nodes
+  const notFloatingNodeIDs = new Set();
+  Object.values(q_graph.edges).forEach((e) => {
+    notFloatingNodeIDs.add(e.subject);
+    notFloatingNodeIDs.add(e.object);
+  });
+
+  // Trim a node if it is floating and marked for deletion
+  const nodesToTrim = Object.keys(q_graph.nodes).filter((id) => (!notFloatingNodeIDs.has(id) && id !== rootNode));
+
+  q_graph.nodes = _.omitBy(q_graph.nodes, (n, id) => nodesToTrim.includes(id));
```

Comment:
Can we combine this with the previous line?

Reply:
Updated in 1630d14212d712dc4715ed737a25f3008a19a59d